keps/sig-cluster-lifecycle/kubeadm/20190916-kubeadm-operator.md

```diff
@@ -0,0 +1,363 @@
+    - [TaskGroupController and Task](#taskgroupcontroller-and-task)
+    - [TaskController](#taskcontroller)
+    - [Observability and Operability](#observability-and-operability)
+      - [Execution order](#execution-order)
+      - [Execution modes](#execution-modes)
+      - [Error management](#error-management)
+    - [Extensibility](#extensibility)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Version Skew Strategy](#version-skew-strategy)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Infrastructure Needed](#infrastructure-needed)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+- [x] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
+- [x] KEP approvers have set the KEP status to `implementable`
+- [x] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+## Summary
+
+Kubeadm operator would like to enable declarative control of kubeadm workflows, automating the execution and the orchestration of such tasks across existing nodes in a cluster.
+
+## Motivation
+
+Kubeadm binary can execute actions only on the machine where it is running e.g. it is not possible to execute actions on other nodes, to copy files across nodes, etc.
+
+As a consequence, most of the kubeadm workflows, like e.g. [kubeadm upgrade](https://kubernetes.io/docs/reference/setup-tools/kubeadm/kubeadm-upgrade/), consists of a complex sequence of tasks that should be manually executed/orchestrated across all the existing nodes in the cluster.
+
+Such a user experience is not ideal, potentially error-prone, and this approach can be considered also a limitation/blocker for implementing more complex workflows like e.g. rotating certificate authorities, change the settings of an existing cluster etc.
+
+This KEP aims to address such problems by applying the operator pattern to kubeadm workflows. 
+
+### Goals
+
+- To allow declarative control of kubeadm action that leads to "in place" mutations[1] of kubeadm generated artifacts. More specifically this initially targets includes kubeadm workflows:
+  - kubeadm upgrade
+  - certificate renewal
+  - change settings/configuration in an existing cluster (NEW) 
+  - certificate authority rotation (NEW)
```

Comment:
Could move this up to be next to certificate renewal in an attempt to group workflows?

Reply:
done!